e2e/suite_test.go

```diff
@@ -25,93 +24,83 @@ func Test_All(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	if suiteConfig.testsToRun != nil {
-		tests := []string{}
-		for testName := range suiteConfig.testsToRun {
-			if _, ok := cases[testName]; !ok {
-				t.Fatalf("unrecognized E2E test case: %q", testName)
-			} else {
-				tests = append(tests, testName)
-			}
-		}
-		t.Logf("Will run the following Agentbaker E2E tests: %s", strings.Join(tests, ", "))
-	} else {
-		t.Logf("Running all Agentbaker E2E tests...")
-	}
+	scenarioTable := scenario.InitScenarioTable(t, suiteConfig.scenariosToRun)
 
 	cloud, err := newAzureClient(suiteConfig.subscription)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	if err := setupCluster(ctx, t, cloud, suiteConfig.location, suiteConfig.resourceGroupName, suiteConfig.clusterName); err != nil {
+	if err := ensureResourceGroup(ctx, t, cloud, suiteConfig.resourceGroupName); err != nil {
 		t.Fatal(err)
 	}
 
-	subnetID, err := getClusterSubnetID(ctx, cloud, suiteConfig.location, suiteConfig.resourceGroupName, suiteConfig.clusterName)
+	clusters, err := listClusters(ctx, t, cloud, suiteConfig.resourceGroupName)
 	if err != nil {
 		t.Fatal(err)
 	}
 
-	kube, err := getClusterKubeClient(ctx, cloud, suiteConfig)
-	if err != nil {
-		t.Fatal(err)
-	}
+	for _, scenario := range scenarioTable {
+		scenario := scenario
+
+		chosenCluster := chooseCompatibleCluster(scenario, clusters)
```

Comment:
hmm. we have a lot of setup inside this loop now -- the execution of each case blocks on the creation of the cluster, and beginning execution of the later cases blocks on cluster creation for all previous steps.

can we simplify or extract the setup somehow?

we could move it inside the actual t.Run(), at risk of causing conflicts, but we could filter out 409s.

Reply:
separate to that, wonder if we can wrap some of this into a setup function to keep the top level loop cleaner

Reply:
definitely agree with wrapping setup into a separate function, will play with moving these bits around to improve parallelization

Reply:
have made some progress on the parallelization stuff, moved that to separate branch which I'll follow up with soon (mainly to keep this PR manageable)